iothub_client/src/blob.c

```diff
@@ -131,257 +135,263 @@ BLOB_RESULT Blob_UploadBlock(
+            LogInfo("Upload to blob has been aborted by the user");
+            uploadOneMoreBlock = 0;
+            result = BLOB_ABORTED;
+        }
+        else if (source == NULL || size == 0)
+        {
+            /*Codes_SRS_BLOB_99_002: [ If the size of the block returned by `getDataCallbackEx` is 0 or if the data is NULL, then `Blob_UploadMultipleBlocksFromSasUri` shall exit the loop. ]*/
+            uploadOneMoreBlock = 0;
+            result = BLOB_OK;
+        }
+        else
+        {
+            if (size > BLOCK_SIZE)
+            {
+                /*Codes_SRS_BLOB_99_001: [ If the size of the block returned by `getDataCallbackEx` is bigger than 4MB, then `Blob_UploadMultipleBlocksFromSasUri` shall fail and return `BLOB_INVALID_ARG`. ]*/
+                LogError("tried to upload block of size %lu, max allowed size is %d", (unsigned long)size, BLOCK_SIZE);
+                result = BLOB_INVALID_ARG;
+                isError = 1;
+            }
+            else if (blockID >= MAX_BLOCK_COUNT)
+            {
+                /*Codes_SRS_BLOB_99_003: [ If `getDataCallbackEx` returns more than 50000 blocks, then `Blob_UploadMultipleBlocksFromSasUri` shall fail and return `BLOB_INVALID_ARG`. ]*/
+                LogError("unable to upload more than %lu blocks in one blob", (unsigned long)MAX_BLOCK_COUNT);
+                result = BLOB_INVALID_ARG;
+                isError = 1;
+            }
+            else
+            {
+                /*Codes_SRS_BLOB_02_023: [ Blob_UploadMultipleBlocksFromSasUri shall create a BUFFER_HANDLE from source and size parameters. ]*/
+                BUFFER_HANDLE requestContent = BUFFER_create(source, size);
+                if (requestContent == NULL)
+                {
+                    /*Codes_SRS_BLOB_02_033: [ If any previous operation that doesn't have an explicit failure description fails then Blob_UploadMultipleBlocksFromSasUri shall fail and return BLOB_ERROR ]*/
+                    LogError("unable to BUFFER_create");
+                    result = BLOB_ERROR;
+                    isError = 1;
+                }
+                else
+                {
+                    result = Blob_UploadBlock(
+                            httpApiExHandle,
+                            relativePath,
+                            requestContent,
+                            blockID,
+                            blockIDList,
+                            httpStatus,
+                            httpResponse);
+
+                    BUFFER_delete(requestContent);
+                }
+
+                /*Codes_SRS_BLOB_02_026: [ Otherwise, if HTTP response code is >=300 then Blob_UploadMultipleBlocksFromSasUri shall succeed and return BLOB_OK. ]*/
+                if (result != BLOB_OK || *httpStatus >= 300)
```

Comment:
Just to make sure, I think it would be safer for us to make sure we check if httpStatus is within the 2xx range. 
LSS, we should consider httpStatus < 200 also an error, since we do not have code to handle 1xx.
LSShorter, 

```suggestion
                if (result != BLOB_OK || *httpStatus < 200 || *httpStatus >= 300)
```

Reply:
I'd actually rather not make this change, or at least not without more research with the Hub and potentially Storage teams which would be a separate issue.  I don't want any functional changes in this (other than initializing the values to NULL, which would prevent a crash in degenerate caes).

Reply:
Per offline discussion, I'd rather keep the refactor isolated over any sort of functional changes even if this is righteous.  Resolving for now.